### PR TITLE
add preprocessor to handle comment directives in sprocket 3.7.0+ , fixes #580

### LINKED ISSUE #580: v.1.8.1 assets precompile //=require will not be transformed
Help us help you! Please choose one:

- [ ] My app crashes with `react-rails`, so I've included the stack trace and the exact steps which make it crash.
- [ ] My app doesn't crash, but I'm getting unexpected behavior. So, I've described the unexpected behavior and suggested a new behavior.
- [ ] I'm trying to use `react-rails` with another library, but I'm having trouble. I've described my JavaScript management setup (eg, Sprockets, Webpack...), how I'm trying to use this other library, and why it's not working.
- [X] I have another issue to discuss.

-------

Hi. We have troubles with the newest release v1.8.1
In our .jsx files we do `//= require ./xyz` which gets transformed into the code from the required file in development, but not when I ran `rake assets:precompile`.

We are using `sprockets 3.7.0` and I am quite sure that the wrong behaviour is a result of following change:
https://github.com/reactjs/react-rails/compare/v1.8.0...v1.8.1#diff-ffad3c61836f28360e5c21dce8362d9fL104 (https://github.com/reactjs/react-rails/commit/99eb4e155458e6a9dc5dc08f6db2b68c1faa611e)

I've locked react-rails to v1.8.0 which resolves the problem, but I don't want to stack to this release.

### PR DESCRIPTION
**Pull request by seuros**
can you please check this PR @forelabs ?

### CODE REVIEW

**Comment by forelabs** [comment]
[X] no comment in compiled asset
[X] comment converted into required component code

works like a charm!

**Comment by rmosolgo** [comment]
@seuros what do you think about adding a test case for this? 

I'm trying to see if I understand ... since we're defining a new mime type in `react-rails`, we also have to tell Sprockets  to preprocess `//= require`s in this new mime type, right? It seems like we should add a test case to cover `//= require`s in the new mime type. 

Maybe you could add a new directory, `assets/javascripts/require_test/`, and make a few files, then add a `.jsx` file which requires them, eg 

```js 
// app/assets/javascripts/require_test/jsx_require_test.jsx
//= require ./jsx_require_child_1
//= require ./jsx_require_child_2

<div className="own-div" />
```

Then, fetch the "compiled" asset and make sure it includes all required content:

```ruby 
jsx_require_asset = Rails.application.assets["require_test/jsx_require_test"].to_s 
# Make sure the asset has required content & own content:
assert_includes jsx_require_asset, jsx_require_child_1_content
assert_includes jsx_require_asset, jsx_require_child_w_content
assert_includes jsx_require_asset, own_content 
```

Or, another way to make sure that this works the way we expect in the future?

**Comment by seuros** [comment]
@rmosolgo Sure thing.

File: test/react/jsx/jsx_prepocessor_test.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+require 'test_helper'
+
+when_sprockets_available do
+  class JSXPreprocessorTest < ActionDispatch::IntegrationTest
+    EXPECTED_JS = <<javascript
+React.createElement(\"div\", { className: \"le-javascript-child\" });
+React.createElement(\"div\", { className: \"le-javascript\" });
+javascript
+
+    test 'accepts harmony: true option' do
```
Comment (rmosolgo) [inline test/react/jsx/jsx_prepocessor_test.rb]: should be `test "executes //= require directives" do` ?

File: test/react/jsx/jsx_prepocessor_test.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+      get '/assets/require_test/jsx_preprocessor_test.js'
+      assert_response :success
+      assert_match(EXPECTED_JS, @response.body)
```
Comment (rmosolgo) [inline test/react/jsx/jsx_prepocessor_test.rb]: seems like this is failing on whitespace :weary: 

maybe you could move this custom assertion to `test_helper` and use it here? https://github.com/reactjs/react-rails/blob/master/test/react/jsx_test.rb#L71-L77

Comment (seuros) [inline test/react/jsx/jsx_prepocessor_test.rb]: Oh, i didn't see it. doing that now.

**Comment by seuros** [comment]
@rmosolgo we can remove the work in progress in sprocket 4 now.

**Comment by rmosolgo** [comment]
🎉  thanks, i'll merge and ship 1.8.2 pronto.

**Comment by rmosolgo** [comment]
removed the sprockets 4 note in 5049473

**Comment by rmosolgo** [comment]
:ship: 1.8.2  🎊  thanks again!

**Pull request merged**